Toolbar custom content

martinpovolny opened the pull request:
To be rebased after #8079 is merged.

martinpovolny commented:
ping @zeari

zeari commented:
Nice! Can I try this with `shared/topology_header` ?

zeari commented:
Added `custom_content('custom', :partial => 'shared/_toolbar_test' )` to the bottom of `app/helpers/application_helper/toolbar/ems_container_center.rb` and got the following exception:

```
[----] F, [2016-05-16T18:14:45.594631 #10554:3a89f14] FATAL -- : Error caught: [ActionView::Template::Error] undefined method `render_to_string' for #<#<Class:0x007f8a562887b8>:0x007f8a562826b0>
/home/azellner/Projects/manageiq/app/helpers/application_helper/toolbar/custom.rb:3:in `render'
/home/azellner/Projects/manageiq/app/helpers/application_helper/toolbar_builder.rb:224:in `block in build'
/home/azellner/Projects/manageiq/app/helpers/application_helper/toolbar_builder.rb:212:in `each'
/home/azellner/Projects/manageiq/app/helpers/application_helper/toolbar_builder.rb:212:in `each_with_index'
/home/azellner/Projects/manageiq/app/helpers/application_helper/toolbar_builder.rb:212:in `build'
/home/azellner/Projects/manageiq/app/helpers/application_helper/toolbar_builder.rb:41:in `build_toolbar'
/home/azellner/Projects/manageiq/app/helpers/application_helper/toolbar_builder.rb:4:in `call'
/home/azellner/Projects/manageiq/app/helpers/application_helper.rb:345:in `build_toolbar'
/home/azellner/Projects/manageiq/app/helpers/toolbar_helper.rb:39:in `block (2 levels) in render_toolbars'

```

martinpovolny commented:
@zeari : there's an issue with `view_context` that differs between the POST and GET request. I wanted to pass the `view_context` to get access to controller variables but I dropped from the code for now and replaced the rendering https://evilmartians.com/chronicles/new-feature-in-rails-5-render-views-outside-of-actions

Please, retry. I wonder if you need access to some controller variables in the partial. I'd prefer not to allow that but I don't know how complex is the stuff you are working on.

zeari commented:
Almost there! Can you help me figure out why its going down a line?

![screencapture-localhost-3000-ems_container-1-1464015069283](https://cloud.githubusercontent.com/assets/11256940/15474793/237f4866-210f-11e6-8b6c-3087ed265681.png)

I replaced the contents of `app/helpers/application_helper/toolbar/ems_container_center.rb` with 
```Ruby
class ApplicationHelper::Toolbar::EmsContainerCenter < ApplicationHelper::Toolbar::Basic
  custom_content('custom', :partial => "shared/topology_header")
end
```

martinpovolny commented:
@zeari: show me your content of the file so that I can check that.

Going to remove the example to make this PR mergeable.

zeari commented:
@martinpovolny its `app/views/shared/_topology_header.html.haml` that all the topology views in the system use.

martinpovolny commented:
@zeari: so this is just some markup issue. 

I know little about that markup, but I'd start by wrapping the content into a `form-group` like the rest of the toolbar stuff is

martinpovolny commented:
@zeari : looking at your example, you are adding a whole new toolbar. But in the example I show here I add just a form-group to a toolbar.

zeari commented:
> @zeari : looking at your example, you are adding a whole new toolbar. But in the example I show here I add just a form-group to a toolbar.

Yeah, seems like the header would need some changes. Overall this change seems to work well.

martinpovolny commented:
@zeari : ok, I am removing the example to make this mergeable. If we decide that we need a whole toolbar as the context of the custom content rather than a group, we can change it.

zeari commented:
> @zeari : ok, I am removing the example to make this mergeable. If we decide that we need a whole toolbar as the context of the custom content rather than a group, we can change it.

:+1:

martinpovolny commented:
Removed example, rebased. Ping @himdel

himdel commented:
LGTM, it seems equivalent, and doesn't seem to break in the UI.. :+1:

zeari commented:
:+1: 
This is needed for for #9156
